src/compute/src/compute_state.rs

```diff
@@ -644,24 +660,196 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
+pub struct PersistPeek {
+    pub(crate) peek: Peek,
+    /// A background task that's responsible for producing the peek results.
+    /// If we're no longer interested in the results, we abort the task.
+    _abort_handle: AbortOnDropAbortHandle,
+    /// The result of the background task, eventually.
+    result: oneshot::Receiver<(PeekResponse, Duration)>,
+    /// The `tracing::Span` tracking this peek's operation
+    span: tracing::Span,
+}
+
+impl PersistPeek {
+    async fn do_peek(
+        persist_clients: &PersistClientCache,
+        metadata: CollectionMetadata,
+        as_of: Timestamp,
+        mfp_plan: SafeMfpPlan,
+        mut limit_remaining: usize,
+    ) -> Result<Vec<(Row, NonZeroUsize)>, String> {
+        let client = persist_clients
+            .open(metadata.persist_location)
+            .await
+            .map_err(|e| e.to_string())?;
+
+        let mut reader: ReadHandle<SourceData, (), Timestamp, Diff> = client
+            .open_leased_reader(
+                metadata.data_shard,
+                Arc::new(metadata.relation_desc),
+                Arc::new(UnitSchema),
+                Diagnostics::from_purpose("persist::peek"),
+            )
+            .await
+            .map_err(|e| e.to_string())?;
+
+        let mut cursor = reader
+            .snapshot_cursor(Antichain::from_elem(as_of))
+            .await
+            .map_err(|since| {
+                format!("attempted to peek at {as_of}, but the since has advanced to {since:?}")
+            })?;
+
+        // Re-used state for processing and building rows.
+        let mut result = vec![];
+        let mut datum_vec = DatumVec::new();
+        let mut row_builder = Row::default();
+        let arena = RowArena::new();
+
+        while limit_remaining > 0 {
+            let Some(batch) = cursor.next().await else {
+                break;
+            };
+            for ((k, v), _, d) in batch {
+                let row = k?.0.map_err(|e| e.to_string())?;
+                let () = v?;
+                let count: usize = d.try_into().map_err(|_| {
+                    format!(
+                        "Invalid data in source, saw retractions ({}) for row that does not exist",
+                        d * -1,
+                    )
+                })?;
+                let Some(count) = NonZeroUsize::new(count) else {
+                    continue;
+                };
+                let mut datum_local = datum_vec.borrow_with(&row);
+                let eval_result = mfp_plan
+                    .evaluate_into(&mut datum_local, &arena, &mut row_builder)
+                    .map_err(|e| e.to_string())?;
+                if let Some(row) = eval_result {
+                    result.push((row, count));
+                    limit_remaining = limit_remaining.saturating_sub(count.get());
+                    if limit_remaining == 0 {
+                        break;
+                    }
+                }
+            }
+        }
```

Comment:
Can you explain why this works? My mental model was that a persist collection consists of multiple batches and one batch could have retractions for updates inserted by a different batch, which means we'd have to read all batches to know which updates are actually still there. Which clearly is a wrong mental model :)

Reply:
> My mental model was that a persist collection consists of multiple batches and one batch could have retractions for updates inserted by a different batch [...]

This is right!

> [...] which means we'd have to read all batches to know which updates are actually still there.

This is not _quite_ right. The main relevant thing to know is that batches are made up of sorted runs of part-files, and we track the smallest key in each part in our metadata. So we have a few tricks available if we want to read just "the first few keys":
- We should only need the first part in each sorted run.
- Since we have stats on the keys, we often only need the first part from a couple different runs - the ones with the smallest keys.

Handwaving a lot, but hopefully that's enough intuition for why we can do better! (The gory details of how parts are fetched and streaming-consolidated together are in `iter.rs`.)